renku/models/provenance/activities.py

```diff
@@ -100,6 +102,16 @@ class Activity(CommitMixin):
         kw_only=True,
     )
 
+    agent = jsonld.ib(context='prov:agent', kw_only=True)
+
+    @agent.default
+    def default_agent(self):
+        """Set the default agent."""
+        return SoftwareAgent(
```

Comment:
Allocating activity is already pretty expensive operation and this adds a bit more to the problem. I propose we create this instantitation as a singleton and then we associate the same singleton object to the newly created Activity, and not allocate new instance every time we create new Activity.

Reply:
good point - I have no idea how to implement a singleton with `attr` though?